clients/geth/specular/rollup/services/base.go

```diff
@@ -33,12 +35,31 @@ type BaseService struct {
 	Wg     sync.WaitGroup
 }
 
+func connectL1(ctx context.Context, endpoint string) (*ethclient.Client, error) {
+	var l1 *ethclient.Client
+	var err error
+	retryOpts := []retry.Option{
+		retry.Context(ctx),
+		retry.Attempts(3),
+		retry.Delay(5 * time.Second),
+		retry.LastErrorOnly(true),
+		retry.OnRetry(func(n uint, err error) {
+			log.Error("Failed to connect to L1", "endpoint", endpoint, "attempt", n, "err", err)
+		}),
+	}
+	err = retry.Do(func() error {
+		l1, err = ethclient.DialContext(ctx, endpoint)
+		return err
+	}, retryOpts...)
+	return l1, err
```

Comment:
nit: think the convention we should try to maintain is to wrap all errors w/ additional context before returning
```suggestion
if err != nil {
    return nil, fmt.Errorf("Could not connect to L1, err: %w", err)
}
return l1, nil
```